Bug: 使用 --reload 加载项无法正常重新加载

**描述问题：**

使用 `nb run --reload` 命令启动的 NoneBot，在文件变更时无法自动重新加载，会引发 **不同** 的错误直接退出进程

**如何复现？**

1. 使用 nb-cli 创建新NoneBot项目
2. 使用 `nb run --reload` 命令启动NoneBot
3. 在src文件夹中写入任意`.py`后缀文件
4. 此时便会抛出Traceback，无法重新加载NoneBot2

由于不清楚是nb-cli的问题还是nonebot2的问题，~~所以先来这里开个issue了~~

***之前在`2.0.0-beta2`版本，通过`.env`文件中的`fastapi_reload`配置项设置的自动重新加载是正常的**

***这个环境运行我写的一些 FastAPI 项目，使用`uvicorn main:app --reload`命令启动时能正常重新加载的**


**环境信息：**

 - OS: Windows 11 x64 22H2
 - Python Version: 3.10.5
 - Nonebot Version: 2.0.0rc3
 - nb-cli Version: 1.0.5
 - Fastapi Version: 0.88.0
 - uvicorn Version: 0.20.0

**截图或日志**

**经过测试，每次自动重新加载抛出的Traeback位置都不完全一致，以下是观察到的几种在重新加载期间抛出的Traceback，可能不是全部的情况（不同Traceback期间项目没有任何改动）**

**第一种**
```python traceback
Watchfiles detected changes in "C:\Users\OREO\Documents\Nonebot\KyaruDev\src\plugins\issue_demo.py". Reloading...
03-21 10:38:15 [INFO] uvicorn | Shutting down
03-21 10:38:15 [WARNING] nonebot | OneBot V11 | WebSocket for Bot 639396094 closed by peer
03-21 10:38:15 [INFO] websockets | connection closed
03-21 10:38:15 [INFO] uvicorn | Finished server process [18156]
03-21 10:38:15 [ERROR] uvicorn | Traceback (most recent call last):
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 674, in lifespan
    await receive()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\lifespan\on.py", line 137, in receive
    return await self.receive_queue.get()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\asyncio\queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

Restarted process [18276].
Shutting down process [18276]...
Traceback (most recent call last):
  File "C:\Users\OREO\Documents\Nonebot\KyaruDev\bot.py", line 4, in <module>
    import nonebot
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\nonebot\__init__.py", line 42, in <module>
    from importlib.metadata import version
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\__init__.py", line 17, in <module>
    from . import _adapters, _meta
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 879, in exec_module
  File "<frozen importlib._bootstrap_external>", line 975, in get_code
  File "<frozen importlib._bootstrap_external>", line 1073, in get_data
KeyboardInterrupt
Stopped reloader.
```

**第二种**
```python traceback
Watchfiles detected changes in "C:\Users\OREO\Documents\Nonebot\KyaruDev\src\plugins\issue_demo.py". Reloading...
03-21 10:38:53 [INFO] uvicorn | Shutting down
03-21 10:38:53 [WARNING] nonebot | OneBot V11 | WebSocket for Bot 639396094 closed by peer
03-21 10:38:53 [INFO] websockets | connection closed
03-21 10:38:53 [INFO] uvicorn | Finished server process [16508]
03-21 10:38:53 [ERROR] uvicorn | Traceback (most recent call last):
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 674, in lifespan
    await receive()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\lifespan\on.py", line 137, in receive
    return await self.receive_queue.get()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\asyncio\queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

Restarted process [17728].
Shutting down process [17728]...
Traceback (most recent call last):
  File "C:\Users\OREO\Documents\Nonebot\KyaruDev\bot.py", line 4, in <module>
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\nonebot\__init__.py", line 42, in <module>
    from importlib.metadata import version
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\__init__.py", line 17, in <module>
    from . import _adapters, _meta
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\_adapters.py", line 3, in <module>
    import email.message
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\email\message.py", line 15, in <module>
    from email import utils
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\email\utils.py", line 30, in <module>
    import datetime
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 879, in exec_module
  File "<frozen importlib._bootstrap_external>", line 1012, in get_code
  File "<frozen importlib._bootstrap_external>", line 672, in _compile_bytecode
KeyboardInterrupt
Stopped reloader.
```

**第三种**
```python traceback
Watchfiles detected changes in "C:\Users\OREO\Documents\Nonebot\KyaruDev\src\plugins\issue_demo.py". Reloading...
03-21 10:39:44 [INFO] uvicorn | Shutting down
03-21 10:39:44 [WARNING] nonebot | OneBot V11 | WebSocket for Bot 639396094 closed by peer
03-21 10:39:44 [INFO] websockets | connection closed
03-21 10:39:44 [INFO] uvicorn | Finished server process [5412]
03-21 10:39:44 [ERROR] uvicorn | Traceback (most recent call last):
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 674, in lifespan
    await receive()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\lifespan\on.py", line 137, in receive
    return await self.receive_queue.get()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\asyncio\queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

Restarted process [17600].
Shutting down process [17600]...
Traceback (most recent call last):
  File "C:\Users\OREO\Documents\Nonebot\KyaruDev\bot.py", line 4, in <module>
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\nonebot\__init__.py", line 42, in <module>
    from importlib.metadata import version
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\__init__.py", line 124, in <module>
    class EntryPoint(
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\__init__.py", line 163, in EntryPoint
    dist: Optional['Distribution'] = None
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 309, in inner
    return cached(*args, **kwds)
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 403, in __getitem__
    return self._getitem(self, parameters)
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 529, in Optional
    arg = _type_check(parameters, f"{self} requires a single type.")
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 164, in _type_check
    arg = _type_convert(arg, module=module, allow_special_forms=allow_special_forms)
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 142, in _type_convert
    return ForwardRef(arg, module=module, is_class=allow_special_forms)
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 668, in __init__
    code = compile(arg, '<string>', 'eval')
KeyboardInterrupt
Stopped reloader.
```

**第四种**
```python traceback
Watchfiles detected changes in "C:\Users\OREO\Documents\Nonebot\test\src\plugins\test.py". Reloading...
03-21 10:41:40 [INFO] uvicorn | Shutting down
03-21 10:41:40 [INFO] uvicorn | Finished server process [13284]
03-21 10:41:40 [ERROR] uvicorn | Traceback (most recent call last):
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 674, in lifespan
    await receive()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\lifespan\on.py", line 137, in receive
    return await self.receive_queue.get()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\asyncio\queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

Restarted process [18168].
Shutting down process [18168]...
Traceback (most recent call last):
  File "<string>", line 1, in <module>

Stopped reloader.
```


```python traceback
Watchfiles detected changes in "C:\Users\OREO\Documents\Nonebot\KyaruDev\src\plugins\issue_demo.py". Reloading...
03-21 10:43:30 [INFO] uvicorn | Shutting down
03-21 10:43:30 [WARNING] nonebot | OneBot V11 | WebSocket for Bot 639396094 closed by peer
03-21 10:43:30 [INFO] websockets | connection closed
03-21 10:43:30 [INFO] uvicorn | Finished server process [13880]
03-21 10:43:30 [ERROR] uvicorn | Traceback (most recent call last):
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\starlette\routing.py", line 674, in lifespan
    await receive()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\uvicorn\lifespan\on.py", line 137, in receive
    return await self.receive_queue.get()
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\asyncio\queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

Restarted process [18388].
Shutting down process [18388]...
Traceback (most recent call last):
  File "C:\Users\OREO\Documents\Nonebot\KyaruDev\bot.py", line 4, in <module>
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\site-packages\nonebot\__init__.py", line 42, in <module>
    from importlib.metadata import version
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\__init__.py", line 17, in <module>
    from . import _adapters, _meta
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\importlib\metadata\_meta.py", line 1, in <module>
    from typing import Any, Dict, Iterator, List, Protocol, TypeVar, Union
  File "C:\Users\OREO\AppData\Local\Programs\Python\Python310\lib\typing.py", line 2264, in <module>
    class NamedTupleMeta(type):
KeyboardInterrupt
Stopped reloader.
```

--- a/nb_cli/handlers/__init__.py
+++ b/nb_cli/handlers/__init__.py
@@ -53,6 +53,7 @@ from .venv import detect_virtualenv as detect_virtualenv
 # isort: split
 
 # signal
+from .signal import shield_signals as shield_signals
 from .signal import remove_signal_handler as remove_signal_handler
 from .signal import install_signal_handler as install_signal_handler
 from .signal import register_signal_handler as register_signal_handler

--- a/nb_cli/handlers/process.py
+++ b/nb_cli/handlers/process.py
@@ -4,12 +4,13 @@ import asyncio
 import subprocess
 from pathlib import Path
 from functools import wraps
+from contextlib import nullcontext
 from typing_extensions import ParamSpec
 from typing import IO, Any, Set, Union, Callable, Optional, Coroutine
 
 from nb_cli.consts import WINDOWS
 
-from .signal import remove_signal_handler, register_signal_handler
+from .signal import shield_signals, remove_signal_handler, register_signal_handler
 
 P = ParamSpec("P")
 
@@ -92,9 +93,12 @@ async def terminate_process(process: asyncio.subprocess.Process) -> None:
     if process.returncode is not None:
         return
 
-    if WINDOWS:
-        os.kill(process.pid, signal.CTRL_C_EVENT)
-    else:
-        process.terminate()
+    context = shield_signals() if WINDOWS else nullcontext()
 
-    await process.wait()
+    with context:
+        if WINDOWS:
+            os.kill(process.pid, signal.CTRL_C_EVENT)
+        else:
+            process.terminate()
+
+        await process.wait()

--- a/nb_cli/handlers/process.py
+++ b/nb_cli/handlers/process.py
@@ -1,6 +1,7 @@
 import os
 import signal
 import asyncio
+import subprocess
 from pathlib import Path
 from functools import wraps
 from typing_extensions import ParamSpec
@@ -65,6 +66,7 @@ async def create_process(
         stdin=stdin,
         stdout=stdout,
         stderr=stderr,
+        creationflags=subprocess.CREATE_NEW_PROCESS_GROUP if WINDOWS else 0,
     )
 
 
@@ -82,6 +84,7 @@ async def create_process_shell(
         stdin=stdin,
         stdout=stdout,
         stderr=stderr,
+        creationflags=subprocess.CREATE_NEW_PROCESS_GROUP if WINDOWS else 0,
     )
 
 

--- a/nb_cli/handlers/process.py
+++ b/nb_cli/handlers/process.py
@@ -97,7 +97,7 @@ async def terminate_process(process: asyncio.subprocess.Process) -> None:
 
     with context:
         if WINDOWS:
-            os.kill(process.pid, signal.CTRL_C_EVENT)
+            os.kill(process.pid, signal.CTRL_BREAK_EVENT)
         else:
             process.terminate()
 

--- a/nb_cli/handlers/signal.py
+++ b/nb_cli/handlers/signal.py
@@ -2,7 +2,9 @@ import signal
 import asyncio
 import threading
 from types import FrameType
-from typing import List, Callable, Optional
+from contextvars import ContextVar
+from contextlib import contextmanager
+from typing import List, Callable, Optional, Generator
 
 HANDLED_SIGNALS = (
     signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
@@ -10,6 +12,8 @@ HANDLED_SIGNALS = (
 )
 
 handlers: List[Callable[[int, Optional[FrameType]], None]] = []
+
+shield_context: ContextVar[bool] = ContextVar("shield_context", default=False)
 
 
 def install_signal_handler() -> None:
@@ -29,6 +33,9 @@ def install_signal_handler() -> None:
 
 
 def handle_signal(signum: int, frame: Optional[FrameType]) -> None:
+    if shield_context.get():
+        return
+
     for handler in handlers:
         handler(signum, frame)
 
@@ -41,3 +48,12 @@ def register_signal_handler(
 
 def remove_signal_handler(handler: Callable[[int, Optional[FrameType]], None]) -> None:
     handlers.remove(handler)
+
+
+@contextmanager
+def shield_signals() -> Generator[None, None, None]:
+    token = shield_context.set(True)
+    try:
+        yield
+    finally:
+        shield_context.reset(token)

--- a/nb_cli/handlers/signal.py
+++ b/nb_cli/handlers/signal.py
@@ -9,11 +9,27 @@ from typing import List, Callable, Optional, Generator
 HANDLED_SIGNALS = (
     signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
     signal.SIGTERM,  # Unix signal 15. Sent by `kill <pid>`.
+    signal.SIGBREAK,  # Windows signal 21. Sent by Ctrl+Break.
 )
 
 handlers: List[Callable[[int, Optional[FrameType]], None]] = []
 
-shield_context: ContextVar[bool] = ContextVar("shield_context", default=False)
+
+class _ShieldContext:
+    def __init__(self) -> None:
+        self._counter = 0
+
+    def acquire(self) -> None:
+        self._counter += 1
+
+    def release(self) -> None:
+        self._counter -= 1
+
+    def active(self) -> bool:
+        return self._counter > 0
+
+
+shield_context = _ShieldContext()
 
 
 def install_signal_handler() -> None:
@@ -33,7 +49,7 @@ def install_signal_handler() -> None:
 
 
 def handle_signal(signum: int, frame: Optional[FrameType]) -> None:
-    if shield_context.get():
+    if shield_context.active():
         return
 
     for handler in handlers:
@@ -52,8 +68,8 @@ def remove_signal_handler(handler: Callable[[int, Optional[FrameType]], None]) -
 
 @contextmanager
 def shield_signals() -> Generator[None, None, None]:
-    token = shield_context.set(True)
+    shield_context.acquire()
     try:
         yield
     finally:
-        shield_context.reset(token)
+        shield_context.release()

--- a/nb_cli/handlers/signal.py
+++ b/nb_cli/handlers/signal.py
@@ -2,15 +2,17 @@ import signal
 import asyncio
 import threading
 from types import FrameType
-from contextvars import ContextVar
 from contextlib import contextmanager
 from typing import List, Callable, Optional, Generator
+
+from nb_cli.consts import WINDOWS
 
 HANDLED_SIGNALS = (
     signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
     signal.SIGTERM,  # Unix signal 15. Sent by `kill <pid>`.
-    signal.SIGBREAK,  # Windows signal 21. Sent by Ctrl+Break.
 )
+if WINDOWS:
+    HANDLED_SIGNALS += (signal.SIGBREAK,)  # Windows signal 21. Sent by Ctrl+Break.
 
 handlers: List[Callable[[int, Optional[FrameType]], None]] = []
 

